### DISCUSSION TITLE
safari使用本地webfont加载非系统字体

### DISCUSSION THREAD

**Discussion by abmcar** [F9y4ng/GreasyFork-Scripts#260] [经验分享(Share Experience)]
safari由于一些安全性问题，无法加载非系统字体
参考链接 https://zhuanlan.zhihu.com/p/81515884

因此想使用自己下载的字体需要通过webfont的形式曲线救国
这里分享一下我自己使用的方法
1.首先搭建本地http服务器
建立一个文件夹，把字体放进去，之后新建py文件
复制以下py代码
``` python
import http.server
import ssl

# 创建一个SSL上下文
context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
context.load_cert_chain(certfile='./server.pem')

# 创建HTTP服务器
httpd = http.server.HTTPServer(('localhost', 8000), http.server.SimpleHTTPRequestHandler)

# 使用SSLContext.wrap_socket()方法包装HTTP服务器的socket
httpd.socket = context.wrap_socket(httpd.socket, server_side=True)

# 启动HTTP服务器
httpd.serve_forever()

```
代码保存完先别运行，还需要https的证书
证书生成命令
```
# 1.生成私钥文件（.key）
openssl genpkey -algorithm RSA -out server.key

# 2.生成证书签名请求文件（.csr）
# 这里可能需要填很多信息，除了Common Name外直接回车不填就行
# Common Name 这个填成 localhost
openssl req -new -key server.key -out server.csr

# 3.生成自签名证书文件（.crt）
openssl x509 -req -in server.csr -signkey server.key -out server.crt

# 4.生成HTTPS服务器.pem文件
cat server.key server.crt > server.pem

```
经过如此操作，我们应该能得到如下文件
![image.png](https://photo-1259569881.cos.ap-nanjing.myqcloud.com/img/20230914152444.png)
运行https.py即可

2.引入safari自定样式表
![image.png](https://photo-1259569881.cos.ap-nanjing.myqcloud.com/img/20230914152549.png)

这里随便找一个位置，新建一个css文件
使用webfont形式引入字体即可
其中「LXGWWenKaiGBScreen」换成自己想引入的字体即可
```
@font-face {
    font-family: 'LXGWWenKaiGBScreen';
    src: url('https://localhost:8000/LXGWWenKaiGBScreen.ttf') format('truetype');
    font-weight: normal;
    font-style: normal;
}		
```
最后效果如图
![image.png](https://photo-1259569881.cos.ap-nanjing.myqcloud.com/img/20230914153602.png)

**Comment by F9y4ng** [comment]
确实是曲线救国的过场。🤣

虽然客户端操作成本稍微有点高，但也算是无奈之举了。之前有用户提出想通过远程字体实现 #232 ，但CSP这道坎还是跨不过呢。

有需要的朋友可以尝试。
